packages/react-server-cli/src/compileClient.js

```diff
@@ -0,0 +1,197 @@
+				absolutePathToPage,
+			];
+		}
+	}
+
+	// now rewrite the routes file out in a webpack-compatible way.
+	const serverRoutes = writeWebpackCompatibleRoutesFile(routes, routesDir, workingDirAbsolute, outputUrl, false);
+	writeWebpackCompatibleRoutesFile(routes, routesDir, workingDirAbsolute, outputUrl, true);
+
+	// finally, let's pack this up with webpack.
+	return {
+		serverRoutes,
+		compiler: webpack(packageCodeForBrowser(entrypoints, outputDirAbsolute, outputUrl, hot, minify)),
+	};
+}
+
+const packageCodeForBrowser = (entrypoints, outputDir, outputUrl, hot, minify) => {
+	const extractTextLoader = require.resolve("./NonCachingExtractTextLoader") + "?{remove:true}!css-loader";
+	let webpackConfig = {
+		entry: entrypoints,
+		output: {
+			path: outputDir,
+			publicPath: outputUrl,
+			filename: "[name].bundle.js",
+			chunkFilename: "[id].bundle.js",
+		},
+		module: {
+			loaders: [
+				{
+					test: /\.jsx?$/,
+					loader: "babel",
+					exclude: /node_modules/,
+				},
+				{
+					test: /.css$/,
+					loader: extractTextLoader,
+					exclude: /node_modules/,
+				},
+			],
+		},
+		plugins: [
+			new ExtractTextPlugin("[name].css"),
+		],
+	};
+
+	if (minify) {
+		webpackConfig.plugins = [
+			...webpackConfig.plugins,
+			new webpack.DefinePlugin({
+				'process.env': {NODE_ENV: '"production"'},
+			}),
+			// TODO: should this be done as babel plugin?
+			new webpack.optimize.UglifyJsPlugin(),
+		];
+	} else {
+		webpackConfig.devtool = "#cheap-module-eval-source-map";
+	}
+
+	if (hot) {
+		webpackConfig.module.loaders.unshift({
+			test: /\.jsx?$/,
+			loader: "react-hot",
+			exclude: /node_modules/,
+		});
+		webpackConfig.plugins = [
+			...webpackConfig.plugins,
+			new webpack.HotModuleReplacementPlugin(),
+			new webpack.NoErrorsPlugin(),
+		];
+	}
+
+	return webpackConfig;
+};
+
+// writes out a routes file that can be used at runtime.
+const writeWebpackCompatibleRoutesFile = (routes, routesDir, workingDirAbsolute, staticUrl, isClient) => {
+	let routesOutput = [];
+
+	const existingMiddleware = routes.middleware ? routes.middleware.map((middlewareRelativePath) => {
+		return `unwrapEs6Module(require("${path.relative(workingDirAbsolute, path.resolve(routesDir, middlewareRelativePath))}"))`
+	}) : [];
+	routesOutput.push(`
+function unwrapEs6Module(module) { return module.__esModule ? module.default : module }
+var coreJsMiddleware = unwrapEs6Module(require('react-server-cli/target/coreJsMiddleware'));
+var coreCssMiddleware = unwrapEs6Module(require('react-server-cli/target/coreCssMiddleware'));
+module.exports = {
+	middleware:[
+		coreJsMiddleware(${JSON.stringify(staticUrl)}),
+		coreCssMiddleware(${JSON.stringify(staticUrl)}),
+		${existingMiddleware.join(",")}
+	],
+	routes:{`);
+
+	for (let routeName in routes.routes) {
+		if (routes.routes.hasOwnProperty(routeName)) {
+			let route = routes.routes[routeName];
+			var relativePathToPage = path.relative(workingDirAbsolute, path.resolve(routesDir, route.page));
```

Comment:
OK, so I __think__ I understand it, although for some reason `reduce` always scrambles my already addled brain a bit :cry: . 

My plan: allow `page` in the routes file to be a string or an object of strings. If it's a single string, I will compile it to:

```
page: {
 default: {
  done: {
   function(cb) {
    cb(require(THE_MODULE_STRING)); // or require.ensure for client-side
   }
  }
}
```
and if it's an object of strings, it'll look just like that but with multiple entries named appropriately instead of a single entry named `default`. I'll work on this now.